enhancements/network/routed-ingress-support-for-primary-udn-attached-vms-with-static-ips.md

```diff
@@ -0,0 +1,810 @@
+required MAC address. We would need MTV to somehow also figure out what IP
+addresses are on the aforementioned interfaces.
+
+To allow OpenShift Virtualization to import VMs - preserving their MAC and IP
+addresses - we need to have both addresses stored somehow in the system (*at
+least* for the VMs which will be migrated into OpenShift Virt).
+
+The simplest alternative (which doesn't require any API changes) is to have MTV
+template the OpenShift Virt VM with the required information - the MAC and IP
+addresses for the primary UDN attachment.
+
+The MAC address is currently defined in the VM spec, it is an attribute of the
+interfaces in the `spec.template.spec.domain.devices.interfaces` array.
+Unfortunately, there isn't a place in the KubeVirt API to request the IP
+addresses for the interface (or network attachment). Given we're already past
+the 4.20 KubeVirt feature freeze, we propose templating the VM with an
+annotation where the interface's IPs will be defined. This annotation
+will be KubeVirt specific, and will be consumed by the KubeVirt
+`ipam-extensions` project, which will mutate the launcher pod to customize the
+cluster default network attachment, by using the multus default network
+feature. The multus default network annotation's value is a
+network-selection-element, which we can use to instruct the pod's IPs and MAC
+addresses. The annotation on the VMs will look like:
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: vm-server
+  namespace: blue
+spec:
+  runStrategy: Always
+  template:
+    metadata:
+      annotations:
+        network.kubevirt.io/addresses: '{"iface1": ["192.168.0.1/24", "fd23:3214::123/64"]}'
+```
+
+When KubeVirt sees this VM, it will create a running VMI with the template's
+spec defined; this the following VMI will be created from the template:
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineInstance
+metadata:
+  name: vm-server
+  namespace: blue
+  annotations:
+    network.kubevirt.io/addresses: '{"iface1": ["192.168.0.1/24", "fd23:3214::123/64"]}'
+spec:
+...
+```
+
+When virt-controller templates the pod for holding this VM it will carry over
+the multus default-network annotation and define that in the launcher pod.
+
+The `ipam-extensions` mutating web-hook will intercept the launcher pod
+creation, and mutate the multus default network annotation: it will decorate
+the network selection element with the ipam-claim-reference attribute, thus
+pointing to the IPAMClaim that is to be used to persist the VM's IP addresses
+when the VMs are shutdown or migrated. This annotation would look like:
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: pod-example
+  annotations:
+    v1.multus-cni.io/default-network: '[{
+      "name": "default",
+      "namespace": "openshift-ovn-kubernetes",
+      "mac": "02:03:04:05:06:07",
+      "ipam-claim-reference": "myvm.isolated-net",  # added by the mutating webhook
+      "ips": [
+        "10.0.0.5/24"
```

Comment:
in order to match the output of the example VM from above, the `ips` should be `"192.168.0.1/24", "fd23:3214::123/64"` no?